operator/controllers/kyma_controller.go

```diff
@@ -121,283 +122,282 @@ func (r *KymaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
+	statusUpdateRequiredFromSync, err := r.SyncConditionsWithModuleStates(ctx, kyma, modules)
+	if err != nil {
+		return err
 	}
 
 	// set ready condition if applicable
 	if kyma.AreAllReadyConditionsSetForKyma() {
 		message := fmt.Sprintf("reconciliation of %s finished!", kyma.Name)
 		logger.Info(message)
 		r.Event(kyma, "Normal", "ReconciliationSuccess", message)
-		return r.updateKymaStatus(ctx, kyma, operatorv1alpha1.KymaStateReady, message)
+		return status.Helper(r).UpdateStatus(ctx, kyma, operatorv1alpha1.KymaStateReady, message)
+	}
+
+	// if the ready condition is not applicable, but we changed the conditions, we still need to issue an update
+	if statusUpdateRequiredFromCreation || statusUpdateRequiredFromSync {
+		return status.Helper(r).UpdateStatus(ctx, kyma, kyma.Status.State, "updating component conditions")
 	}
 
 	return nil
 }
 
-func (r *KymaReconciler) HandleDeletingState(ctx context.Context, logger *logr.Logger, kyma *operatorv1alpha1.Kyma) (bool, error) {
+func (r *KymaReconciler) HandleDeletingState(ctx context.Context, kyma *operatorv1alpha1.Kyma) (bool, error) {
+	logger := log.FromContext(ctx)
 	templates, err := release.GetTemplates(ctx, r, kyma)
 	if err != nil {
-		return false, fmt.Errorf("deletion cannot proceed - templates could not be fetched: %w", err)
+		return false, status.Helper(r).UpdateStatus(ctx, kyma, operatorv1alpha1.KymaStateError, "templates could not be fetched")
 	}
 
-	for _, component := range kyma.Spec.Components {
-		actualComponentStruct, err := util.GetUnstructuredComponentFromTemplate(templates, component.Name, kyma)
-		if err != nil {
-			return false, err
-		}
-
-		if err = r.Get(ctx, client.ObjectKeyFromObject(actualComponentStruct), actualComponentStruct); err == nil {
+	modules, err := util.ParseTemplates(kyma, templates)
+	if err != nil {
+		return false, err
+	}
+	for name, module := range modules {
+		if err = r.Get(ctx, client.ObjectKeyFromObject(module.Unstructured), module.Unstructured); err == nil {
 			// component CR still exists
-
 			logger.Info(fmt.Sprintf("deletion cannot proceed - waiting for component CR %s to be deleted for %s",
-				actualComponentStruct.GetName(), client.ObjectKeyFromObject(kyma)))
+				name, client.ObjectKeyFromObject(kyma)))
 			return true, nil
 		} else if !errors.IsNotFound(err) {
 			// unknown error while getting component CR
 			return false, fmt.Errorf("deletion cannot proceed - unknown error: %w", err)
 		}
 	}
 
+	// remove finalizer
 	logger.Info("All component CRs have been removed, removing finalizer",
 		"resource", client.ObjectKeyFromObject(kyma))
-
 	if err := remote.RemoveFinalizerFromRemoteKyma(ctx, r, client.ObjectKeyFromObject(kyma)); client.IgnoreNotFound(err) != nil {
 		return false, err
 	}
 	logger.Info("removed remote finalizer",
 		"resource", client.ObjectKeyFromObject(kyma))
-
-	// remove finalizer
 	controllerutil.RemoveFinalizer(kyma, labels.Finalizer)
```

Comment:
Assume remote cluster is gone, and the client received not found error, and we ignored this error, which means we can remove the finalizer of kyma (local). 
Is it possible in this case, in the above `for name, module := range modules` loop, we enter into modules error state which prevent this remote unreachable kyma to be deleted.
What I means is should we introduce some logic if the remote kyma is not found, we force remove kyma and all related modules without error check.

Reply:
The logic means that if the object can not be found remotely, it will not try to remove the finalizer as it considers it already deleted and it will behave like normal. So your logic is actually covered (I think)